src/v/cluster/self_test/netcheck.cc

```diff
@@ -0,0 +1,166 @@
+        return run_benchmark_fiber(
+          run_fiber_opts(max_timeout, _opts.duration), peer, m);
+    });
+    m.set_total_time(ss::lowres_clock::now() - now);
+    auto result = m.to_st_result();
+    result.name = _opts.name;
+    result.info = fmt::format("Test performed against node: {}", peer);
+    if (_cancelled) {
+        result.warning = "Run was manually canclled";
+    }
+    co_return result;
+}
+
+ss::future<> netcheck::run_benchmark_fiber(
+  run_fiber_opts fiber_state, model::node_id peer, metrics& m) {
+    /// run_fiber_opts manages the the RPC request timeouts to ensure that
+    /// the maximum amount of time for an individual RPC never exceeds the
+    /// remaining time of the test
+    auto now = ss::lowres_clock::now();
+    while (now < fiber_state.test_end() && !_cancelled) {
+        if (unlikely(_as.abort_requested())) {
+            throw netcheck_aborted_exception();
+        }
+        /// Measure the time it takes to complete a full round trip
+        co_await m.measure([this, peer, &fiber_state]() mutable {
+            netcheck_request r{
+              .source = _self, .buf = create_sample_blob(_opts.request_size)};
+            return _connections.local()
+              .with_node_client<self_test_rpc_client_protocol>(
+                _self,
+                ss::this_shard_id(),
+                peer,
+                10s,
+                [&fiber_state,
+                 r = std::move(r)](self_test_rpc_client_protocol c) mutable {
+                    /// This RPC is not really 'cancellable' without shutting
+                    /// down the connection, however these requests should
+                    /// complete very quickly, so no explicit shutdown mechanism
+                    /// (other then the _cancelled bool) exists
+                    return c.netcheck(
+                      std::move(r), rpc::client_opts(fiber_state.test_end()));
+                })
+              .then([this, &fiber_state](auto reply) {
+                  return process_netcheck_reply(std::move(reply), fiber_state);
+              });
+        });
+        now = ss::lowres_clock::now();
+    }
+}
+
+ss::future<size_t> netcheck::process_netcheck_reply(
+  result<rpc::client_context<netcheck_response>> reply,
+  run_fiber_opts& fiber_state) {
+    if (!reply) {
+        if (reply.error() == rpc::errc::client_request_timeout) {
+            /// timeout expired, let above loop continue
+            co_return -1;
+        }
+        throw network_benchmark_exception(reply.error().message());
+    } else if (reply.value().data.buf.size_bytes() == 0) {
+        /// Server is already processing requests from a benchmark
+        /// via another node, sleep until can proceed, measurement
+        /// will be thrown away
+        if (fiber_state.not_started()) {
+            /// still waiting for server to be available
+            co_await ss::sleep(200ms);
```

Comment:
Non-abortable sleeps are undesirable (even when short), is there an abort source we could pass in?